src/newgrf_badge_gui.cpp

```diff
@@ -245,3 +246,205 @@ std::unique_ptr<DropDownListItem> MakeDropDownListBadgeIconItem(const std::share
+	}
+
+	void Draw(const Rect &full, const Rect &r, bool sel, int click_result, Colours bg_colour) const override
+	{
+		bool rtl = (_current_text_dir == TD_RTL);
+		int w = SETTING_BUTTON_WIDTH;
+
+		int state = 0;
+		if (sel && click_result != 0) {
+			if (click_result == this->click_up) state = 1;
+			if (click_result == this->click_down) state = 2;
+		}
+
+		Rect br = r.WithWidth(w, TEnd ^ rtl).CentreTo(w, SETTING_BUTTON_HEIGHT);
+		DrawArrowButtons(br.left, br.top, this->button_colour, state, this->click_up, this->click_down);
+
+		this->TBase::Draw(full, r.Indent(w + WidgetDimensions::scaled.hsep_wide, TEnd ^ rtl), sel, click_result, bg_colour);
+	}
+
+private:
+	int click_up; ///< Click result for up button. Button is inactive if 0.
+	int click_down; ///< Click result for down button. Button is inactive if 0.
+	Colours button_colour; ///< Colour of buttons.
+};
+
+using DropDownListToggleMoverItem = DropDownMover<DropDownToggle<DropDownString<DropDownListItem>>>;
+
+static constexpr int CLICK_MOVE_UP = 1;
+static constexpr int CLICK_MOVE_DOWN = 2;
+static constexpr int CLICK_TOGGLE_ICON = 3;
+
+DropDownList BuildBadgeClassConfigurationList(const GUIBadgeClasses &gui_classes, uint columns, std::span<const StringID> column_separators)
+{
+	DropDownList list;
+
+	list.push_back(MakeDropDownListStringItem(STR_BADGE_CONFIG_RESET, INT_MAX));
+	if (gui_classes.GetClasses().empty()) return list;
+	list.push_back(MakeDropDownListDividerItem());
+
+	const BadgeClassID front = gui_classes.GetClasses().front().class_index;
+	const BadgeClassID back = gui_classes.GetClasses().back().class_index;
+
+	for (uint i = 0; i < columns; ++i) {
+		for (const auto &gc : gui_classes.GetClasses()) {
+			if (gc.column_group != i) continue;
+
+			bool first = (i == 0 && gc.class_index == front);
+			bool last = (i == columns - 1 && gc.class_index == back);
+			list.push_back(std::make_unique<DropDownListToggleMoverItem>(first ? 0 : CLICK_MOVE_UP, last ? 0 : CLICK_MOVE_DOWN, COLOUR_YELLOW, gc.visible, CLICK_TOGGLE_ICON, COLOUR_YELLOW, COLOUR_GREY, GetString(GetClassBadge(gc.class_index)->name), gc.class_index.base()));
+		}
+
+		if (i >= column_separators.size()) continue;
+
+		if (column_separators[i] == STR_NULL) {
+			list.push_back(MakeDropDownListDividerItem());
+		} else {
+			list.push_back(MakeDropDownListStringItem(column_separators[i], INT_MIN + i, false, true));
+		}
+	}
+
+	return list;
+}
+
+/**
+ * Toggle badge class visibility.
+ * @param feature Feature being used.
+ * @param class_badge Class badge.
+ * @param click Dropdown click reuslt.
+ */
+static void BadgeClassToggleVisibility(GrfSpecFeature feature, Badge &class_badge, int click_result)
+{
+	auto config = GetBadgeClassConfiguration(feature);
+	auto it = std::ranges::find(config, class_badge.label, &BadgeClassConfigItem::label);
+	if (it == std::end(config)) return;
+
+	if (click_result == CLICK_TOGGLE_ICON) it->show_icon = !it->show_icon;
```

Comment:
Why is this check here, and not in the `switch` in `HandleBadgeConfigurationDropDownClick`?